File: DriveMate/DriveMate/Services/ResponseParser.swift
import Foundation

struct ParsedResponse {
    let cleanText: String
    let searchQuery: String?
    let callNumber: String?
}

enum ResponseParser {

    private static let searchRegex = try! NSRegularExpression(pattern: "\\[BUSCAR:(.+?)\\]")
    private static let callRegex = try! NSRegularExpression(pattern: "\\[LLAMAR:(.+?)\\]")

    static func parse(_ text: String) -> ParsedResponse {
        let nsText = text as NSString
        let range = NSRange(location: 0, length: nsText.length)

        var searchQuery: String?
        var callNumber: String?

        if let match = searchRegex.firstMatch(in: text, range: range) {
            searchQuery = nsText.substring(with: match.range(at: 1))
                .trimmingCharacters(in: .whitespaces)
        }

        if let match = callRegex.firstMatch(in: text, range: range) {
            callNumber = nsText.substring(with: match.range(at: 1))
                .trimmingCharacters(in: .whitespaces)
        }

        // Remove tags from visible text
        var clean = searchRegex.stringByReplacingMatches(
            in: text, range: range, withTemplate: ""
        )
        let cleanNS = clean as NSString
        clean = callRegex.stringByReplacingMatches(
            in: clean, range: NSRange(location: 0, length: cleanNS.length), withTemplate: ""
        )
        clean = clean.trimmingCharacters(in: .whitespacesAndNewlines)

        return ParsedResponse(
            cleanText: clean,
            searchQuery: searchQuery,
            callNumber: callNumber
        )
    }
}

File: DriveMateAndroid/app/src/main/java/com/drivemate/service/ResponseParser.kt
package com.drivemate.service

data class ParsedResponse(
    val cleanText: String,
    val searchQuery: String?,
    val callNumber: String?
)

object ResponseParser {

    private val searchPattern = Regex("""\[BUSCAR:(.+?)]""")
    private val callPattern = Regex("""\[LLAMAR:(.+?)]""")

    fun parse(text: String): ParsedResponse {
        val searchQuery = searchPattern.find(text)?.groupValues?.get(1)?.trim()
        val callNumber = callPattern.find(text)?.groupValues?.get(1)?.trim()

        var clean = text
        clean = searchPattern.replace(clean, "")
        clean = callPattern.replace(clean, "")
        clean = clean.trim()

        return ParsedResponse(
            cleanText = clean,
            searchQuery = searchQuery,
            callNumber = callNumber
        )
    }
}
